src/main/java/com/metamx/metrics/JvmMonitor.java

```diff
@@ -206,14 +115,165 @@ public boolean doMonitor(ServiceEmitter emitter)
       emitter.emit(builder.build("jvm/bufferpool/used", pool.getMemoryUsed()));
       emitter.emit(builder.build("jvm/bufferpool/count", pool.getCount()));
     }
+  }
 
-    return true;
+  private void emitGcMetrics(ServiceEmitter emitter)
+  {
+    gcCounters.emit(emitter, dimensions);
   }
 
-  private void emitGcMetrics(GcCounters gcCounters, ServiceEmitter emitter)
+  /*
+   * The following GC-related code is partially based on
+   * https://github.com/aragozin/jvm-tools/blob/e0e37692648951440aa1a4ea5046261cb360df70/
+   * sjk-core/src/main/java/org/gridkit/jvmtool/PerfCounterGcCpuUsageMonitor.java
+   */
+  private class GcCounters
+  {
+    private List<GcGeneration> generations = new ArrayList<>();
+
+    GcCounters()
+    {
+      long currentProcessId = SigarUtil.getCurrentProcessId();
+      // connect to itself
+      JStatData jStatData = JStatData.connect(currentProcessId);
+      Map<String, JStatData.Counter<?>> jStatCounters = jStatData.getAllCounters();
+
+      generations.add(new GcGeneration(jStatCounters, 0, "young"));
+      generations.add(new GcGeneration(jStatCounters, 1, "old"));
+      // Removed in Java 8 but still actual for previous Java versions
+      if (jStatCounters.containsKey("sun.gc.generation.2.name")) {
+        generations.add(new GcGeneration(jStatCounters, 2, "perm"));
+      }
+    }
+
+    void emit(ServiceEmitter emitter, Map<String, String[]> dimensions)
+    {
+      for (GcGeneration generation : generations) {
+        final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+        MonitorUtils.addDimensionsToBuilder(builder, dimensions);
+        generation.emit(emitter, builder);
+      }
+    }
+  }
+
+  private class GcGeneration
```

Comment:
Currently this class could be `static`, I think? However, if `emit()` method is changed so that it creates new builders for each emitted event (see a review comment in this method), it should fill those builders with `dimensions`, i. e. access a field of the outer class, then `GcGeneration` couldn't be `static`.